File: Leetcode/buy-and-sell-stocks.rb
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/  

def max_profit(prices)
  results = []  
  for i in 0...prices.length-1
    # puts buy = {date: i, price: prices[i]}         
    # newPrices = prices.dup
    # newPrices.slice!(i)    
    for j in i..prices.length-1
      profit = prices[j] - prices[i]
      if  profit >0
        results.push(profit)
      end      
      
      # puts "Day of buying: " + i.to_s
      # puts "Price of buying: " + prices[i].to_s
      # puts "Day of selling: " + j.to_s
      # puts "Price of selling: " + prices[j].to_s
      # puts "Profit: " + (prices[j] - prices[i] ).to_s
    end
    # p newPrices

    # profit.push(prices[i+1]-prices[i])    
    # newprices    
    # profit.push(prices[i+1])
    
  end 
  # puts "Results: " + results.to_s
  # p "Results: " + results.to_s
  # p "Length: " + results.length.to_s
  return 0 if results.length == 0
  results.max
end


  
def test input, expected
  puts "Begin test"
  # puts "Begin test with imput: " + input
  result = max_profit(input)
  if(result === expected)
    puts "Test passed."
  else
    puts "Input: " + input.to_s
    puts "Expected: " + expected.to_s
    puts "Got: " + result.to_s
    puts "Test failed."      
  end    
  puts "------"
end

test [7,1,5,3,6,4], 5
test [7,6,4,3,1], 0

File: Leetcode/buy-and-sell-stocks.js
//https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
var maxProfit = function(prices) {
  let results = []
  for (i = 0; i < prices.length ; i++ ){      
      for (j = i; j < prices.length; j++ ){
        profit = prices[j] - prices[i]
        if(profit > 0){
          results.push(profit)
        }
        // console.log(i, "day of buying");
        // console.log(prices[i], "buying price");
        // console.log(j, "day of selling")
        // console.log(prices[j], "selling price")
        // console.log(profit, "Profit")
        // console.log( "------")
      }
  }
  // console.log(results, "results")
  if( results.length === 0){
    return 0;
  }
  return Math.max(...results)  
};

function test(input, expected){
  console.log("Begin test")
  //console.log( "Begin test with imput: " + input)
  let result = maxProfit(input)
  if(result === expected){
    console.log( "Test passed.")
  } else{
    console.log( "Input: " + input)
    console.log( "Expected: " + expected)
    console.log( "Got: " + result)
    console.log( "Test failed." )
  }
  console.log( "------")
}

test ([7,1,5,3,6,4], 5)
// test ([7,6,4,3,1], 0)